In [1]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
import numpy as np
from sklearn.datasets import load_digits
import matplotlib.pyplot as plt
digits = load_digits()
from sklearn.model_selection import train_test_split

### Cross-Validation
### It is a technique used to assess how well our Machine learning models perform on unseen data 
### We can get Generalized model (Low Bias and Low Variance)

In [2]:
X_train, X_test, y_train, y_test = train_test_split(digits.data,digits.target,test_size=0.3)

In [3]:
lr = LogisticRegression(solver='liblinear',multi_class='ovr')
lr.fit(X_train, y_train)
lr.score(X_test, y_test)

0.9722222222222222

In [4]:
svm = SVC(gamma='auto')
svm.fit(X_train, y_train)
svm.score(X_test, y_test)

0.32592592592592595

In [5]:
rf = RandomForestClassifier(n_estimators=40)
rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.9833333333333333

### CV

In [7]:
from sklearn.model_selection import cross_val_score
score_lr=cross_val_score(LogisticRegression(solver='liblinear',multi_class='ovr'), digits.data, digits.target,cv=3)
print(score_lr)
print("Avg :",np.average(score_lr))

[0.89482471 0.95325543 0.90984975]
Avg : 0.9193099610461881


In [10]:
score_svm =cross_val_score(SVC(gamma='auto'), digits.data, digits.target,cv=3)
print(score_svm)
print('Avg :',np.average(score_svm))

[0.38063439 0.41068447 0.51252087]
Avg : 0.4346132442960489


In [12]:
score_rf=cross_val_score(RandomForestClassifier(n_estimators=40),digits.data, digits.target,cv=3)
print(score_rf)
print('Avg :',np.average(score_rf))

[0.92654424 0.92988314 0.92821369]
Avg : 0.9282136894824707


### Before K Fold apply	After K Fold applied (Avg)
### Logistic Regression	97%	91%
### SVM	32%	43% 
### Random Forest	98%	92%

#### Based on the above table, we will go with Random Forest for this dataset for production. But we have to monitor the model performance based on the data drift and as the business case changes, we have to revisit the model and redeploy.